storage-proofs-update/src/vanilla.rs

```diff
@@ -0,0 +1,1205 @@
+        );
+
+        // Setup read-only mmaps for sector_key_path and staged_data_path inputs.
+        let sector_key_data = mmap_read(sector_key_path)?;
+        let staged_data = mmap_read(staged_data_path)?;
+
+        // Setup writable mmap for new_replica_path output.
+        let mut new_replica_data = mmap_write(new_replica_path)?;
+
+        // Re-open staged_data as Data (type)
+        let mut new_data = Data::from_path(staged_data_path.to_path_buf());
+        new_data.ensure_data()?;
+
+        // Generate tree_d over the staged_data.
+        let tree_d = StackedDrg::<TreeR, TreeDHasher>::build_binary_tree::<TreeDHasher>(
+            new_data.as_ref(),
+            tree_d_new_config,
+        )?;
+
+        let comm_d_new = tree_d.root();
+
+        let comm_r_old = <TreeRHasher as Hasher>::Function::hash2(&comm_c, &comm_r_last_old);
+        let phi = phi(&comm_d_new, &comm_r_old);
+
+        let end = staged_data_path_metadata.len() as u64;
+
+        // chunk_size is the number of Fr elements to process in parallel chunks.
+        let chunk_size: usize = std::cmp::min(base_tree_nodes_count, CHUNK_SIZE_MIN);
+
+        // data_block_size is the segment length that we're processing
+        // in Fr elements (i.e. chunk_size * sizeof(Fr)).
+        let data_block_size: usize = chunk_size * FR_SIZE;
+
+        // Right-shift each node-index by `get_high_bits_shr` to get its `h` high bits.
+        let node_index_bit_len = nodes_count.trailing_zeros() as usize;
+        let get_high_bits_shr = node_index_bit_len - h;
+
+        // Precompute all possible `2^h` number of rho values.
+        let rhos: Vec<Fr> = (0..1 << h)
+            .map(|high| {
+                let high: TreeRDomain = Fr::from(high as u64).into();
+                <TreeRHasher as Hasher>::Function::hash2(&phi, &high).into()
+            })
+            .collect();
+
+        Vec::from_iter((0..end).step_by(data_block_size))
+            .into_par_iter()
+            .zip(new_replica_data.par_chunks_mut(data_block_size))
+            .try_for_each(|(chunk_index, replica_data)| -> Result<()> {
+                for i in (0..data_block_size as u64).step_by(FR_SIZE) {
+                    let input_index = (chunk_index as usize) + i as usize;
+                    let output_index = i as usize;
+
+                    // Get the `h` high bits from the node-index.
+                    let node_index = input_index / FR_SIZE;
+                    let high = node_index >> get_high_bits_shr;
+                    let rho = rhos[high];
+
+                    let sector_key_fr =
+                        bytes_into_fr(&sector_key_data[input_index..input_index + FR_SIZE])?;
+                    let staged_data_fr =
+                        bytes_into_fr(&staged_data[input_index..input_index + FR_SIZE])?;
+
+                    let new_replica_fr = sector_key_fr + (staged_data_fr * rho);
+                    let new_replica_bytes = fr_into_bytes(&new_replica_fr);
```

Comment:
could avoid this allocation by having a version of `fr_into_bytes` that writes directly into a slice